instance/api/filters.py

```diff
@@ -65,3 +66,116 @@ def get_filtered_queryset(self, queryset, organization):
         Return filtered queryset by organization.
         """
         return queryset.filter(owner=organization)
+
+
+class InstanceFilterBackend(filters.BaseFilterBackend):
+    """
+    More complex filter for Instance that allows users to filter on fields.
+
+    Currently allowed fields:
+    - deployment_type
+    - lifecycle
+    - name
+    - notes
+    - openedx_release
+    - status
+    - tag
+    """
+
+    def _filter_name(self, queryset, value):
+        """
+        Filter the InstanceRef queryset on partial name.
+        """
+        if value:
+            return queryset.filter(name__icontains=value)
+        return queryset
+
+    def _filter_notes(self, queryset, value):
+        "Filter the InstanceRef queryset on partial notes."
+        if value:
+            return queryset.filter(notes__icontains=value)
+        return queryset
+
+    def _filter_status(self, queryset, value):
+        """
+        Filter the InstanceRef queryset on exact app server status.
+        """
+        if value:
+            # The .distinct is important, because an instance reference can
+            # have multiple appservers with the same status. In that case
+            # there'll be duplicate rows.
+            return queryset.filter(openedxappserver_set___status=value).distinct()
+        return queryset
+
+    def _filter_openedx_release(self, queryset, value):
+        """
+        Filter the InstanceRef queryset on exact openedx_release.
+        """
+        if value:
+            instances = OpenEdXInstance.objects.filter(openedx_release__iexact=value)
+            return queryset.filter(instance_id__in=instances)
+        return queryset
+
+    def _filter_tag(self, queryset, value):
+        """
+        Filter the InstanceRef queryset on the OpenEdXInstance containing at
+        least one matching tag.
+        """
+        if value:
+            instances = OpenEdXInstance.objects.filter(tags__name__iexact=value)
+            return queryset.filter(instance_id__in=instances)
+        return queryset
+
+    def _filter_deployment_type(self, queryset, value):
+        """
+        Filter the InstanceRef queryset on the matching deployment_type.
+        """
+        if value:
+            return queryset.filter(deployment__type=value)
+        return queryset
+
+    def _filter_lifecycle(self, queryset, value):
+        """
+        Filter the InstanceRef queryset on whether it will be released
+        to production or not.
+
+        Allowed values:
+        - production
+        - sandbox
+        """
+        # There currently isn't a clean way of determining this
+        # so the below is just a hack.
+
+        if not value:
+            return queryset
+
+        # TODO: Find the setting that contains this value.
+        default_prod_monitor_email = ['urgent@opencraft.com']
```

Comment:
@keithgg The relevant setting is `PROD_APPSERVER_FAIL_EMAILS`.

Reply:
Thanks! PR updated.